### PR TITLE
Move signal window functions to ATen; add Blackman window

### PR DESCRIPTION
**Pull request by ssnl**
cc @rafaelvalle .

relevant issue https://github.com/pytorch/pytorch/issues/3775

### CODE REVIEW

File: aten/src/ATen/native/TensorFactories.cpp [hunk 1]
```diff
@@ -321,5 +328,85 @@ Tensor zeros_like(const Tensor& self, const Type& dtype) {
...
+  }
+  if (window_length == 1) {
+    return at::ones(dtype, {1});
+  }
+  if (periodic) {
+    window_length += 1;
+  }
+  auto window = at::arange(dtype, window_length);
+  window.mul_(M_PI * 2. / static_cast<double>(window_length - 1)).cos_().mul_(-beta).add_(alpha);
+  return periodic ? window.narrow(0, 0, window_length - 1) : window;
```
Comment (vishwakftw) [inline aten/src/ATen/native/TensorFactories.cpp]: What does `narrow` do?

File: test/test_torch.py [hunk 2]
```diff
@@ -3897,23 +3897,27 @@ def test_svd(self):
...
-            for periodic in [True, False]:
-                ref = torch.from_numpy(signal.get_window(scipy_name, size, fftbins=periodic))
-                self.assertEqual(torch_method(size, periodic=periodic), ref)
+    def _test_signal_window_functions(self, device='cpu'):
+        def test(name):
+            torch_method = getattr(torch, name + '_window')
+            for size in [1, 2, 5, 10, 50, 100, 1024, 2048]:
+                for periodic in [True, False]:
+                    res = torch_method(size, periodic=periodic, device=device)
+                    ref = torch.from_numpy(signal.get_window(name, size, fftbins=periodic))
```
Comment (vishwakftw) [inline test/test_torch.py]: This is causing test to fail, I think.

File: aten/src/ATen/native/TensorFactories.cpp [hunk 3]
```diff
@@ -321,5 +328,85 @@ Tensor zeros_like(const Tensor& self, const Type& dtype) {
...
+  if (periodic) {
+    window_length += 1;
+  }
+  // from https://en.wikipedia.org/wiki/Window_function#Blackman_window
+  auto window = at::arange(dtype, window_length).mul_(M_PI / static_cast<double>(window_length - 1));
+  window = window.mul(4).cos_().mul_(0.08) - window.mul(2).cos_().mul_(0.5) + 0.42;
+  return periodic ? window.narrow(0, 0, window_length - 1) : window;
+}
+
+Tensor hamming_window(const Type& dtype, int64_t window_length, bool periodic, double alpha, double beta) {
```
[the remark this answers is not in the document]
Comment (SsnL) [reply aten/src/ATen/native/TensorFactories.cpp]: The default values are in `native_functions.yaml`, which are close approximations `alpha=0.54` and `beta=0.46`. These are also the values used by NumPy. :)

[the remark this answers is not in the document]
Comment (vishwakftw) [reply aten/src/ATen/native/TensorFactories.cpp]: Oh, I didn't notice it over there. Sorry.

**Comment by vishwakftw** [comment]
I was looking at the set of window functions supported by SciPy and some functions from Wikipedia, as well. Since you added the Blackman window, I thought there might be more to add.

**Comment by SsnL** [comment]
@vishwakftw yeah I consulted @rafaelvalle on what we should add, and he suggest that we add this one along the others we already have. I don't think that we want to add all window fns. Just these common ones should be sufficient. I'm not an expert in this area though, so feel free to correct me if I'm missing some important window fn!

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -5425,3 +5425,170 @@ def parse_kwargs(desc):
...
+Returns:
+    Tensor: A 1-D tensor of size :math:`(\text{{window_length}},)` containing the window
+
+""".format(**factory_common_args))
+
+
+add_docstr(torch.bartlett_window,
+           """
+bartlett_window(window_length, periodic=True, dtype=None, \
+layout=torch.strided, device=None, requires_grad=False) -> Tensor
```
Comment (zou3519) [inline torch/_torch_docs.py]: is it intentional that you're supporting device and requires_grad? The old functions only supported `dtype`

**Comment by SsnL** [comment]
Yes it’s intentional @zou3519 .

**Review by zou3519 — APPROVED** [review]
lgtm!

**Review by soumith — APPROVED** [review]

**Pull request merged**